Framework/Catalog/src/ONCat.cpp

```diff
@@ -0,0 +1,449 @@
+ *   client must prompt the user to enter their credentials again.
+ *
+ * - There is no way for the client to know whether or not the refresh
+ *   will be successful ahead of time (i.e., whether a token has been
+ *   revoked server-side), since the OAuth spec provides no mechanism to
+ *   check the validity of a refresh token.
+ *
+ * - Tokens can be revoked at any time with absolutely no notice as part
+ *   of standard OAuth practice.  Also, only a limited number of tokens
+ *   can exist for each unique client / user combination at any one
+ *   time.
+ *
+ * Hopefully it is clear that working with OAuth client-side requires
+ * you to use an almost-Pythonic "ask for forgiveness rather than for
+ * permission" strategy -- i.e., code as if locally-stored tokens can be
+ * refreshed, but be ready to prompt the user for their credentials if
+ * the refresh fails.
+ *
+ * Some useful links with related information:
+ *
+ * - http://qr.ae/TUTke2 (quora.com)
+ * - https://stackoverflow.com/a/30826806/778572
+ *
+ * @param true if a user is "logged in", else false.
+ */
+bool ONCat::isUserLoggedIn() const {
+  if (m_flow == OAuthFlow::CLIENT_CREDENTIALS) {
+    // No users are ever authenticated as part of this flow.
+    return false;
+  }
+
+  return m_tokenStore->getToken().is_initialized();
+}
+
+void ONCat::logout() {
+  // Currently, ONCat OAuth does *not* allow clients to revoke tokens
+  // that are no longer needed (though this is defined in the OAtuh
+  // spec).  A "logout", then, is simply throwing away whatever token we
+  // previously stored client-side.
+  m_tokenStore->setToken(boost::none);
+  g_log.debug() << "Logging out." << std::endl;
+  ;
+}
+
+/**
+ * Log in as part of the Resource Ownder Credentials flow so that
+ * authenticated resources may be accessed on behalf of a user.
+ *
+ * @param username : The XCAMS / UCAMS ID of the user.
+ * @param password : The XCAMS / UCAMS password of the user.
+ *
+ * @exception Mantid::Catalog::Exception::InvalidCredentialsError :
+ *   Thrown when the given credentials are not valid.
+ */
+void ONCat::login(const std::string &username, const std::string &password) {
+  if (m_flow != OAuthFlow::RESOURCE_OWNER_CREDENTIALS) {
+    g_log.warning()
+        << "Unexpected usage detected!  "
+        << "Logging in with user credentials in not required (and is not "
+        << "supported) when machine-to-machine credentials are being used."
+        << std::endl;
+    ;
+    return;
+  }
+
+  Poco::Net::HTMLForm form(Poco::Net::HTMLForm::ENCODING_MULTIPART);
+  form.set("username", username);
+  form.set("password", password);
+  form.set("client_id", m_clientId);
+  if (m_clientSecret) {
+    form.set("client_secret", m_clientSecret.get());
+  }
+  form.set("grant_type", "password");
+
+  m_internetHelper->setBody(form);
+
+  try {
+    std::stringstream ss;
+
+    const int statusCode =
+        m_internetHelper->sendRequest(m_url + "/oauth/token", ss);
```

Comment:
Care must be taken that once this is connected to something that isn't a test it is using https.

Reply:
We actually disallow HTTP connections to http://oncat.ornl.gov and redirect to https://oncat.ornl.gov, so we're safe here.